src/censored.jl

```diff
@@ -0,0 +1,409 @@
+            v = prob_lower * abs2(iszero(prob_lower) ? oneunit(lower) : lower - μ) +
+                prob_upper * abs2(iszero(prob_upper) ? oneunit(upper) : upper - μ)
+            return v
+        end
+    end
+
+    dtrunc = _to_truncated(d)
+    μinterval = mean(dtrunc)
+    μ = prob_interval * μinterval
+    if !iszero(prob_lower)
+        μ += prob_lower * lower
+    end
+    if !iszero(prob_upper)
+        μ += prob_upper * upper
+    end
+
+    v = prob_interval * (var(dtrunc) + abs2(μinterval - μ))
+    if !iszero(prob_lower)
+        v += prob_lower * abs2(lower - μ)
+    end
+    if !iszero(prob_upper)
+        v += prob_upper * abs2(upper - μ)
+    end
+    return v
+end
+
+# this expectation also uses the following relation:
+# 𝔼_{x ~ τ}[-log d(x)] = S[τ] - log P_{x ~ d₀}(l ≤ x ≤ u)
+#   + (P_{x ~ d₀}(x = l) (log P_{x ~ d₀}(x = l) - log P_{x ~ d₀}(x ≤ l)) + 
+#      P_{x ~ d₀}(x = u) (log P_{x ~ d₀}(x = u) - log P_{x ~ d₀}(x ≥ u))
+#   ) / P_{x ~ d₀}(l ≤ x ≤ u),
+# where S[τ] is the entropy of τ.
+
+function entropy(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    if lower !== missing
+        prob_lower_inc = cdf(d0, lower)
+        pl = value_support(typeof(d0)) === Discrete ? pdf(d0, lower) : 0
+        prob_lower = prob_lower_inc - pl
+        entropy_lower = -xlogx(prob_lower_inc)
+    else
+        pl = prob_lower = entropy_lower = 0
+    end
+    if upper !== missing
+        prob_upper = ccdf(d0, upper)
+        pu = value_support(typeof(d0)) === Discrete ? pdf(d0, upper) : 0
+        entropy_upper = -xlogx(prob_upper + pu)
+    else
+        pu = prob_upper = entropy_upper = 0
+    end
+    result = entropy_lower + entropy_upper
+    prob_interval = 1 - (prob_lower + prob_upper)
+    # truncation contains no probability
+    iszero(prob_interval) && return result
+
+    dtrunc = _to_truncated(d)
+    result += prob_interval * entropy(dtrunc) - xlogx(prob_interval) + xlogx(pl) + xlogx(pu)
+    return result
+end
+
+
+#### Evaluation
+
+function pdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if lower !== missing && x == lower
+        result = cdf(d0, x)
+        _eqnotmissing(x, upper) ? one(result) : result
+    elseif _eqnotmissing(x, upper)
+        _ccdf_inc(d0, x)
+    else
+        result = pdf(d0, x)
+        _in_open_interval(x, lower, upper) ? result : zero(result)
+    end
+end
+
+function logpdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if lower !== missing && x == lower
+        result = logcdf(d0, x)
+        _eqnotmissing(x, upper) ? zero(result) : result
+    elseif _eqnotmissing(x, upper)
+        _logccdf_inc(d0, x)
+    else
+        result = logpdf(d0, x)
+        _in_open_interval(x, lower, upper) ? result : oftype(result, -Inf)
+    end
+
+end
+
+function loglikelihood(d::Censored, x::AbstractArray{<:Real, M}) where {M}
```

Comment:
I guess this method is added to improve performance? It should work without it but probably less performant?

It seems `M` is not used:
```suggestion
function loglikelihood(d::Censored, x::AbstractArray{<:Real})
```